### PR TITLE
ROC-3551: Add a proper S2S health check

### PR DESCRIPTION
**Pull request by sabahirfan**
In this PR, health check indicator is leasing new S2S token via `/lease` endpoint and then retrieving service name via `/details` endpoint ti identify health of `service-auth-provider-api`

Jira: [ROC-3551](https://tools.hmcts.net/jira/browse/ROC-3551)

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java [hunk 1]
```diff
@@ -16,20 +21,55 @@
...
+    private Health check() {
         try {
             InternalHealth internalHealth = this.serviceAuthorisationHealthApi.health();
-            return new Health.Builder(internalHealth.getStatus()).build();
+            if (!internalHealth.getStatus().getCode().equalsIgnoreCase(Status.UP.getCode())) {
+                return new Health.Builder(internalHealth.getStatus()).build();
+            } else {
+                String token = generateToken();
+                String serviceName = this.serviceAuthorisationApi.getServiceName(token);
+                if (!this.microService.equalsIgnoreCase(serviceName)) {
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java]: seems unnecessary

File: src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java [hunk 2]
```diff
@@ -16,20 +21,55 @@
...
         this.serviceAuthorisationHealthApi = serviceAuthorisationHealthApi;
+        this.secret = secret;
+        this.microService = microService;
+        this.serviceAuthorisationApi = serviceAuthorisationApi;
     }
 
     @Override
     public Health health() {
+        return check();
+    }
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java]: why an extra method?

File: build.gradle [hunk 3]
```diff
@@ -10,7 +10,7 @@ plugins {
 }
 
 group 'uk.gov.hmcts.reform'
-version '1.0.0'
+version '1.0.1'
```
Comment (rkondratowicz) [inline build.gradle]: is this a bug fix?

File: src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java [hunk 4]
```diff
@@ -16,20 +21,55 @@
...
+
+    private Health check() {
         try {
             InternalHealth internalHealth = this.serviceAuthorisationHealthApi.health();
-            return new Health.Builder(internalHealth.getStatus()).build();
+            if (!internalHealth.getStatus().getCode().equalsIgnoreCase(Status.UP.getCode())) {
+                return new Health.Builder(internalHealth.getStatus()).build();
+            } else {
+                String token = generateToken();
+                String serviceName = this.serviceAuthorisationApi.getServiceName(token);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java]: this library is used by services that are not registered in s2s (they only validate incoming service) thus this check cannot be performed.

File: build.gradle (hunk 3 above)
Comment (sabahirfan) [reply build.gradle]: Yes, in production there was a scenario when health check was showing up but the actual requests were not being served.

Comment (nitinprabhu) [reply build.gradle]: could you please more details about this what error did the service returned ?

Comment (nitinprabhu) [reply build.gradle]: you are testing the whole functionality as part of health check which doesn't seem to be ideal.

Comment (sabahirfan) [reply build.gradle]: Please have a look at the story : [ROC-3551](https://tools.hmcts.net/jira/browse/ROC-3551)

File: src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java (hunk 4 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java]: Was not aware of this.

Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java]: @rkondratowicz can you give an example of service that just uses `getServiceName` to validate token?

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java]: draft-store and send-letter

Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/reform/authorisation/healthcheck/ServiceAuthHealthIndicator.java]: Thanks @rkondratowicz.

**Comment by geek-soft** [comment]
This story has been raised as a response to one of the incidents we had in production. That day S2S `/health` was responding with `UP` status while any single call to `/details` used to time out. Finally service got restarted which resolved the issue. We never got to the bottom of the issue.

Idea here was to extend health indicator by adding extra checks that calls service endpoints in the same way as applications do it. At the time the story was created we did not realise that some services that have no underlying dependencies do generate service tokens.

Since that complications we are happy to drop the idea of extending health checks and reproach the problem the next time when we have some S2S related incident.

**Pull request closed without merging**